File: backend/src/common_agent/api/routers/audit.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Annotated, Literal
from uuid import UUID

from fastapi import APIRouter, Query, Request
from pydantic import BaseModel, ConfigDict

from common_agent.api.errors import AppError, ErrorEnvelope
from common_agent.audit import (
    AuditAction,
    AuditEvent,
    AuditOutcome,
    AuditQuery,
    AuditResourceType,
    AuditService,
)
from common_agent.tenancy import TenantAccess, TenantRole

router = APIRouter(
    prefix="/api/v1/audit-events",
    tags=["audit"],
    responses={403: {"model": ErrorEnvelope}, 422: {"model": ErrorEnvelope}},
)


class AuditEventResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sequence: int
    event_id: UUID
    tenant_id: UUID | None
    actor_user_id: UUID | None
    action: AuditAction
    outcome: AuditOutcome
    request_id: UUID
    trace_id: str
    resource_type: AuditResourceType | None
    resource_id: str | None
    error_code: str | None
    occurred_at: datetime
    retention_until: datetime
    previous_hash: str
    event_hash: str


class AuditPageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[AuditEventResponse]
    next_cursor: str | None


class AuditIntegrityResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_count: int
    first_sequence: int | None
    last_sequence: int | None
    last_hash: str
    verified: bool
    broken_sequence: int | None


class AuditPolicyResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    retention_days: int
    max_events_per_scope: int
    automatic_deletion: bool


@router.get("", response_model=AuditPageResponse)
async def list_audit_events(
    request: Request,
    scope: Literal["tenant", "platform"] = "tenant",
    actor_user_id: UUID | None = None,
    resource_type: AuditResourceType | None = None,
    resource_id: Annotated[str | None, Query(min_length=1, max_length=128)] = None,
    action: AuditAction | None = None,
    occurred_from: datetime | None = None,
    occurred_to: datetime | None = None,
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    cursor: Annotated[str | None, Query(min_length=1, max_length=20)] = None,
) -> AuditPageResponse:
    access = _owner_access(request)
    try:
        query = AuditQuery(
            tenant_id=access.tenant_id if scope == "tenant" else None,
            actor_user_id=actor_user_id,
            resource_type=resource_type,
            resource_id=resource_id,
            action=action,
            occurred_from=_utc(occurred_from),
            occurred_to=_utc(occurred_to),
            limit=limit,
            cursor=cursor,
        )
        page = await _service(request).page(query)
    except ValueError as error:
        raise AppError("invalid_audit_query", "审计查询参数不合法", 422, False) from error
    return AuditPageResponse(
        items=[_event_response(event) for event in page.items],
        next_cursor=page.next_cursor,
    )


@router.get("/integrity", response_model=AuditIntegrityResponse)
async def verify_audit_integrity(
    request: Request,
    scope: Literal["tenant", "platform"] = "tenant",
) -> AuditIntegrityResponse:
    access = _owner_access(request)
    tenant_id = access.tenant_id if scope == "tenant" else None
    integrity = await _service(request).verify(tenant_id)
    return AuditIntegrityResponse(
        event_count=integrity.event_count,
        first_sequence=integrity.first_sequence,
        last_sequence=integrity.last_sequence,
        last_hash=integrity.last_hash,
        verified=integrity.verified,
        broken_sequence=integrity.broken_sequence,
    )


@router.get("/policy", response_model=AuditPolicyResponse)
async def audit_policy(request: Request) -> AuditPolicyResponse:
    _owner_access(request)
    policy = _service(request).policy
    return AuditPolicyResponse(
        retention_days=policy.retention_days,
        max_events_per_scope=policy.max_events_per_scope,
        automatic_deletion=policy.automatic_deletion,
    )


def _service(request: Request) -> AuditService:
    service = getattr(request.app.state, "audit", None)
    if not isinstance(service, AuditService):
        raise AppError("audit_unavailable", "审计服务暂时不可用", 503, True)
    return service


def _owner_access(request: Request) -> TenantAccess:
    access = getattr(request.state, "tenant_access", None)
    if not isinstance(access, TenantAccess):
        raise AppError("tenant_access_denied", "无权访问该租户", 403, False)
    if access.role is not TenantRole.OWNER:
        raise AppError("tenant_admin_forbidden", "当前角色无权管理工作区", 403, False)
    return access


def _utc(value: datetime | None) -> datetime | None:
    if value is None:
        return None
    if value.tzinfo is None:
        raise ValueError("audit time must include a timezone")
    return value.astimezone(UTC)


def _event_response(event: AuditEvent) -> AuditEventResponse:
    return AuditEventResponse(
        sequence=event.sequence,
        event_id=event.event_id,
        tenant_id=event.tenant_id,
        actor_user_id=event.actor_user_id,
        action=event.action,
        outcome=event.outcome,
        request_id=event.request_id,
        trace_id=event.trace_id,
        resource_type=event.resource_type,
        resource_id=event.resource_id,
        error_code=event.error_code,
        occurred_at=event.occurred_at,
        retention_until=event.retention_until,
        previous_hash=event.previous_hash,
        event_hash=event.event_hash,
    )


__all__ = ["AuditEventResponse", "router"]

File: frontend/src/api/audit.ts
import { z } from "zod";

import type { components } from "./generated/schema";
import { apiClient } from "./client";
import { toApiClientError } from "./errors";

export const auditActionSchema = z.enum([
  "auth.register",
  "auth.login",
  "auth.logout",
  "auth.recovery.reset",
  "auth.member.provisioned",
  "tenant.created",
  "employee.created",
  "employee.configuration_and_bindings.updated",
  "tool.grants.updated",
  "tool.credentials.updated",
  "mcp.source.created",
  "mcp.source.updated",
  "mcp.source.discovered",
  "tool.capability.created",
  "tool.capability.updated",
  "tool.capabilities.imported",
  "tool.called",
  "model.configuration.created",
  "model.configuration.updated",
  "model.configuration.verified",
  "knowledge.base.created",
  "knowledge.document.uploaded",
  "knowledge.document.retry_started",
  "resource.deleted",
  "conversation.reply.started",
  "workflow.configuration.updated",
  "workflow.run.started",
  "workflow.run.stopped",
  "security.permission.denied",
  "security.request.denied",
]);

export const auditResourceTypeSchema = z.enum([
  "user",
  "session",
  "tenant",
  "employee",
  "model_configuration",
  "mcp_source",
  "tool_capability",
  "knowledge_base",
  "knowledge_document",
  "conversation",
  "workflow",
  "workflow_run",
]);

const digestSchema = z.string().regex(/^[0-9a-f]{64}$/);
const auditEventSchema = z.strictObject({
  sequence: z.number().int().positive(),
  event_id: z.uuid(),
  tenant_id: z.uuid().nullable(),
  actor_user_id: z.uuid().nullable(),
  action: auditActionSchema,
  outcome: z.enum(["started", "succeeded", "denied", "failed"]),
  request_id: z.uuid(),
  trace_id: z.string().regex(/^[0-9a-f]{32}$/),
  resource_type: auditResourceTypeSchema.nullable(),
  resource_id: z.string().min(1).max(128).nullable(),
  error_code: z.string().min(1).max(128).nullable(),
  occurred_at: z.iso.datetime({ offset: true }),
  retention_until: z.iso.datetime({ offset: true }),
  previous_hash: digestSchema,
  event_hash: digestSchema,
});
const auditPageSchema = z.strictObject({
  items: z.array(auditEventSchema),
  next_cursor: z.string().min(1).nullable(),
});
const auditIntegritySchema = z.strictObject({
  event_count: z.number().int().nonnegative(),
  first_sequence: z.number().int().positive().nullable(),
  last_sequence: z.number().int().positive().nullable(),
  last_hash: digestSchema,
  verified: z.boolean(),
  broken_sequence: z.number().int().nonnegative().nullable(),
});
const auditPolicySchema = z.strictObject({
  retention_days: z.number().int().min(30).max(3650),
  max_events_per_scope: z.number().int().min(100).max(10_000_000),
  automatic_deletion: z.literal(false),
});

export type AuditAction = components["schemas"]["AuditAction"];
export type AuditResourceType = components["schemas"]["AuditResourceType"];
export type AuditEvent = components["schemas"]["AuditEventResponse"];
export type AuditPage = components["schemas"]["AuditPageResponse"];
export type AuditIntegrity = components["schemas"]["AuditIntegrityResponse"];
export type AuditPolicy = components["schemas"]["AuditPolicyResponse"];
export type AuditScope = "tenant" | "platform";

export interface AuditQuery {
  scope?: AuditScope;
  actor_user_id?: string;
  resource_type?: AuditResourceType;
  resource_id?: string;
  action?: AuditAction;
  occurred_from?: string;
  occurred_to?: string;
  limit?: number;
  cursor?: string;
}

export async function fetchAuditEvents(query: AuditQuery = {}): Promise<AuditPage> {
  try {
    const response = await apiClient.get<unknown>("/audit-events", {
      params: compactQuery(query),
    });
    return auditPageSchema.parse(response.data);
  } catch (error) {
    throw toApiClientError(error);
  }
}

export async function fetchAuditIntegrity(scope: AuditScope = "tenant"): Promise<AuditIntegrity> {
  try {
    const response = await apiClient.get<unknown>("/audit-events/integrity", {
      params: { scope },
    });
    return auditIntegritySchema.parse(response.data);
  } catch (error) {
    throw toApiClientError(error);
  }
}

export async function fetchAuditPolicy(): Promise<AuditPolicy> {
  try {
    const response = await apiClient.get<unknown>("/audit-events/policy");
    return auditPolicySchema.parse(response.data);
  } catch (error) {
    throw toApiClientError(error);
  }
}

function compactQuery(query: AuditQuery): Record<string, string | number> {
  return Object.fromEntries(
    Object.entries(query).filter((entry): entry is [string, string | number] => {
      const value = entry[1];
      return value !== undefined && value !== "";
    }),
  );
}
